File: 2018/python/advent_24.py
import copy

from common.advent_lib import *
from collections import Counter, defaultdict, deque
import collections as coll
import functools as func
import itertools as it
import operator as oper
import re
import more_itertools as it2
import numpy as np
import sympy as sym

lines = full_input(24)

immune, infection = re.split(r"(?<=\n)\n", lines)
id = 0


def parse_unit(line):
    global id
    id += 1
    units, hp, damage, init = map(int, re.findall(r"\d+", line))
    dmg_type = re.search(r"does \d+ ([a-z]+) damage", line).group(1)
    weaknesses = re.search(r"(?<=weak to )([a-z, ]+)+", line)
    weaknesses = weaknesses.group(1).split(", ") if weaknesses else []
    immunities = re.search(r"(?<=immune to )([a-z, ]+)+", line)
    immunities = immunities.group(1).split(", ") if immunities else []
    return (id, units, hp, tuple(weaknesses), tuple(immunities), damage, dmg_type, init, 0)


PARTY, ID, UNITS, HP, WEAKNESSES, IMMUNITIES, DAMAGE, DAMAGE_TYPE, INITIATIVE, TURNS = 0, 1, 2, 3, 4, 5, 6, 7, 8, 9

immune = [*map(parse_unit, immune.split("\n")[1:-1])]
immune = [("immune", *unit) for unit in immune]
infection = [*map(parse_unit, infection.split("\n")[1:-1])]
infection = [("infection", *unit) for unit in infection]
original_units = [*immune, *infection]


def effective_power(unit):
    return unit[DAMAGE] * unit[UNITS]


def damage_done(attacker, defender):
    if attacker[DAMAGE_TYPE] in defender[IMMUNITIES]:
        return 0
    coeff = 2 if attacker[DAMAGE_TYPE] in defender[WEAKNESSES] else 1
    return coeff * attacker[DAMAGE] * attacker[UNITS]

def play_game(units):
    turns = 0
    while len({unit[0] for unit in units}) > 1:
        targets = dict()
        # target selection
        for current_unit in sorted(units, key=lambda unit: (-effective_power(unit),-unit[INITIATIVE])):
            enemies = [unit for unit in units if unit[PARTY] != current_unit[PARTY]]
            available_targets = [*sorted([enemy for enemy in enemies if damage_done(current_unit, enemy) > 0 and enemy[ID] not in targets.values()],
                                         key=lambda enemy: (-damage_done(current_unit, enemy),
                                                            -effective_power(enemy),
                                                            -enemy[INITIATIVE]))]
            if available_targets:
                targets[current_unit[ID]] = available_targets[0][ID]
        if not targets:
            return "draw",-1
        while [unit for unit in units if unit[TURNS] == turns]:
            current_unit = [*sorted([unit for unit in units if unit[TURNS] == turns], key=lambda unit: -unit[INITIATIVE])][0]
            if current_unit[ID] in targets:
                target_id = targets[current_unit[ID]]
                target = [unit for unit in units if unit[ID]==target_id]
                if target:
                    target = target[0]
                    dealt_damage = damage_done(current_unit, target)
                    units_killed = dealt_damage // target[HP]
                    other_units = [unit for unit in units if unit[ID] != target[ID]]
                    if units_killed < target[UNITS]:
                        target = list(target)
                        target[UNITS] -= units_killed
                        other_units += [tuple(target)]
                    units = other_units
            units = [(*current_unit[:-1], current_unit[TURNS] + 1), *[unit for unit in units if unit[ID] != current_unit[ID]]]

        turns += 1
    return units[0][PARTY] if len({unit[0] for unit in units}) == 1 else "draw", sum(unit[UNITS] for unit in units)

part1 = play_game(copy.deepcopy(original_units))[1]
print("Part 1:", part1)

def boost_unit(unit,boost):
    *rest,dmg,dtype,init,turns=unit
    return *rest,dmg+boost,dtype,init,turns

boost = 1
while True:
    boosted_immune=[*[boost_unit(unit,boost) for unit in copy.deepcopy(immune)], *copy.deepcopy(infection)]
    winner,points= play_game(boosted_immune)
    if winner == 'immune':
        part2 = points
        break
    boost += 1

print("Part 2:", part2)

File: 2018/java/Advent24.java
import java.util.*;
import java.util.regex.*;
import java.util.stream.*;

public class Advent24 extends Advent {

  Pattern pattern = Pattern.compile("(\\d+) units each with (\\d+) hit points (\\((immune|weak) to (\\w+(, )?)+(; (immune|weak) to (\\w+(, )?)+)?\\) )?with an attack that does (\\d+) (\\w+) damage at initiative (\\d+)");

  Pattern weaknessPattern = Pattern.compile("weak to ((\\w+(, )?)+)");
  Pattern immunityPattern = Pattern.compile("immune to ((\\w+(, )?)+)");

  List<Group> groups = new ArrayList<>();

  Comparator<Group> groupComparator = Comparator.comparingInt((Group g) -> g.effectivePower())
                                                .thenComparingInt((Group g) -> g.initiative)
                                                .reversed();

  int boost = 0;

  public Advent24() {
    super(24);
  }

  @Override
  protected void parseInput() {
    groups.clear();
    String faction = null;
    for (String line : input) {
      if (line.endsWith(":")) {
        faction = line.substring(0, line.length() - 1);
      } else if (line.length() == 0) {
        continue;
      } else {
        parseLine(line, faction);
      }
    }
  }

  private void parseLine(String line, String faction) {
    Matcher m = pattern.matcher(line);
    m.find();
    Group group = new Group();
    group.faction = faction;
    group.units = Integer.parseInt(m.group(1));
    group.hp = Integer.parseInt(m.group(2));
    int baseDamage = Integer.parseInt(m.group(11));
    int damage = faction.equals("Immune System") ? baseDamage + boost : baseDamage;
    group.damage = damage;
    group.attackType = m.group(12);
    group.initiative = Integer.parseInt(m.group(13));
    String modifiers = m.group(3);
    group.weaknesses = parseWeaknesses(modifiers);
    group.immunities = parseImmunities(modifiers);
    groups.add(group);
  }

  private List<String> parseWeaknesses(String modifiers) {
    List<String> result = new ArrayList<>();
    if (modifiers == null) return result;
    Matcher m = weaknessPattern.matcher(modifiers);
    if (m.find()) {
      result.addAll(Arrays.asList(m.group(1).split(", ")));
    }
    return result;
  }

  private List<String> parseImmunities(String modifiers) {
    List<String> result = new ArrayList<>();
    if (modifiers == null) return result;
    Matcher m = immunityPattern.matcher(modifiers);
    if (m.find()) {
      result.addAll(Arrays.asList(m.group(1).split(", ")));
    }
    return result;
  }

  @Override
  protected String part1() {
    playGame();
    return "" + winnersAlive();
  }

  private void playGame() {
    Map<Group, Group> targets;
    while (bothFactionsAlive()) {
      targets = findTargets();
      if (targets.isEmpty()) {
        return;
      }
      for (var pair : targets.entrySet()) {
        pair.getKey().attack(pair.getValue());
      }
    }
  }

  private boolean bothFactionsAlive() {
    return infectionCount() > 0 && immuneCount() > 0;
  }

  private long infectionCount() {
    return groups.stream()
                 .filter(g -> g.faction.equals("Infection"))
                 .filter(Group::alive)
                 .count();
  }

  private long immuneCount() {
    return groups.stream()
                 .filter(g -> !g.faction.equals("Infection"))
                 .filter(Group::alive)
                 .count();
  }

  private Map<Group, Group> findTargets() {
    Map<Group, Group> result = new TreeMap<>(Comparator.comparingInt((Group g) -> g.initiative).reversed());
    List<Group> eligibleGroups = groups.stream()
                                       .filter(Group::alive)
                                       .sorted(groupComparator)
                                       .collect(Collectors.toList());
    List<Group> eligibleTargets = new ArrayList<>(eligibleGroups);
    while (!eligibleGroups.isEmpty()) {
      Group group = eligibleGroups.get(0);
      eligibleGroups.remove(0);
      Group target = findTarget(group, eligibleTargets);
      if (target != null) {
        result.put(group, target);
        eligibleTargets.remove(target);
      }
    }
    return result;
  }

  private Group findTarget(Group group, List<Group> targets) {
    return targets.stream()
                  .filter(Group::alive)
                  .filter(g -> !g.faction.equals(group.faction))
                  .filter(g -> group.damageTo(g) > 0)
                  .sorted(damageComparator(group))
                  .findFirst()
                  .orElse(null);
  }

  private Comparator<Group> damageComparator(Group group) {
    return Comparator.comparingInt((Group g) -> group.damageTo(g))
                     .thenComparingInt((Group g) -> g.effectivePower())
                     .thenComparingInt((Group g) -> g.initiative)
                     .reversed();
  }

  private long winnersAlive() {
    return groups.stream()
    .filter(Group::alive)
    .mapToInt(g -> g.units)
    .sum();
  }

  @Override
  protected String part2() {
    while (!winner().equals("Immune System")) {
      ++boost;
      parseInput();
      playGame();
    }
    return "" + winnersAlive();
  }

  private String winner() {
    List<String> factionsAlive = groups.stream()
                                       .filter(Group::alive)
                                       .map(g -> g.faction)
                                       .distinct()
                                       .collect(Collectors.toList());
    return factionsAlive.size() > 1 ? "Deadlock" : factionsAlive.get(0);
  }

  class Group {
    String faction;
    int units;
    int hp;
    int damage;
    int initiative;
    String attackType;
    List<String> weaknesses;
    List<String> immunities;

    boolean alive() {
      return units > 0;
    }

    int damageTo(Group other) {
      if (other.immunities.contains(this.attackType)) return 0;
      int factor = other.weaknesses.contains(this.attackType) ? 2 : 1;
      return this.units*this.damage*factor;
    }

    int effectivePower() {
      return this.units*this.damage;
    }

    void attack(Group other) {
      if (!this.alive()) return;
      int damage = this.damageTo(other);
      int killed = Math.min(damage / other.hp, other.units);
      if (debug) sopl("Killed ", killed);
      other.units -= killed;
    }

    public String toString() {
      StringBuilder sb = new StringBuilder();
      sb.append("faction=").append(faction).append(", ")
        .append("units=").append(units).append(", ")
        .append("hp=").append(hp).append(", ")
        .append("damage=").append(damage).append(", ")
        .append("initiative=").append(initiative).append(", ")
        .append("attackType=").append(attackType).append(", ")
        .append("weaknesses=").append(weaknesses).append(", ")
        .append("immunities=").append(immunities);
      return sb.toString();
    }
  }
}
